x-pack/plugins/security_solution/server/lib/detection_engine/routes/rules/perform_bulk_action_route.ts

```diff
@@ -86,93 +81,54 @@ const normalizeErrorResponse = (errors: RuleActionError[]): NormalizedRuleError[
-    attributes: {
-      errors: normalizeErrorResponse(errors).map(({ message, ...error }) => ({
-        ...error,
-        message:
-          message.length > MAX_ERROR_MESSAGE_LENGTH
-            ? `${message.slice(0, MAX_ERROR_MESSAGE_LENGTH - 3)}...`
-            : message,
-      })),
-      rules: {
-        total: rulesCount,
-        failed: errorsCount,
-        succeeded: rulesCount - errorsCount,
-      },
-    },
+const buildBulkResponse = (
+  response: KibanaResponseFactory,
+  fetchRulesOutcome: PromisePoolOutcome<string, RuleAlertType>,
+  bulkActionOutcome: PromisePoolOutcome<RuleAlertType, RuleAlertType | null>
+) => {
+  const errors = [...fetchRulesOutcome.errors, ...bulkActionOutcome.errors];
+  const summary = {
+    failed: errors.length,
+    succeeded: bulkActionOutcome.results.length,
+    total: bulkActionOutcome.results.length + errors.length,
   };
-};
 
-const executeActionAndHandleErrors = async (
-  rule: Rule,
-  action: RuleActionFn
-): Promise<RuleActionResult> => {
-  try {
-    await action(rule);
-  } catch (err) {
-    const { message, statusCode } = transformError(err);
-    return {
-      error: { message, statusCode },
-      rule: { id: rule.id, name: rule.name },
-    };
-  }
-};
-
-const executeBulkAction = async (rules: Rule[], action: RuleActionFn, abortSignal: AbortSignal) =>
-  initPromisePool({
-    concurrency: MAX_RULES_TO_UPDATE_IN_PARALLEL,
-    items: rules,
-    executor: async (rule) => executeActionAndHandleErrors(rule, action),
-    abortSignal,
-  });
-
-const getRulesByIds = async ({
-  ids,
-  rulesClient,
-  isRuleRegistryEnabled,
-  abortSignal,
-}: {
-  ids: string[];
-  rulesClient: RulesClient;
-  isRuleRegistryEnabled: boolean;
-  abortSignal: AbortSignal;
-}) => {
-  const readRulesExecutor = async (id: string) => {
-    try {
-      const rule = await readRules({ id, rulesClient, isRuleRegistryEnabled, ruleId: undefined });
-      if (rule == null) {
-        throw Error('Can`t fetch a rule');
-      }
-      return { rule };
-    } catch (err) {
-      const { message, statusCode } = transformError(err);
-      return {
-        error: { message, statusCode },
-        rule: { id },
-      };
-    }
+  const results = {
+    updated: bulkActionOutcome.results
+      .filter(({ item, result }) => item.id === result?.id)
+      .map(({ result }) => result && internalRuleToAPIResponse(result)),
+    created: bulkActionOutcome.results
+      .filter(({ item, result }) => result != null && result.id !== item.id)
+      .map(({ result }) => result && internalRuleToAPIResponse(result)),
+    deleted: bulkActionOutcome.results
+      .filter(({ result }) => result == null)
+      .map(({ item }) => internalRuleToAPIResponse(item)),
   };
 
-  const { results } = await initPromisePool({
-    concurrency: MAX_RULES_TO_UPDATE_IN_PARALLEL,
-    items: ids,
-    executor: readRulesExecutor,
-    abortSignal,
-  });
+  if (errors.length > 0) {
+    return response.customError({
```

Comment:
Not sure where it's coming from, but there is another property `error` appeared to be added in error response:

```
attributes: {errors: [{,…}],…}
error: "Internal Server Error"
message: "Bulk edit partially failed"
statusCode: 500

```

`error` property is not needed here, I think we have to remove it, so it won't confuse API's users

Reply:
Agree; I don't think we need to duplicate `statusCode` in the response body. We can find all the places the status code is coming from and clean them up. I think it would be OK to do that in a follow-up PR.

Reply:
I meant property `error`:

```
error: "Internal Server Error"

```

and then we have property message:

```
message: "Bulk edit partially failed"
```

Not sure we need error here as message already describes reason of failure

Reply:
The `error` field is added by the Kibana core server and is not configurable.

Reply:
Is this coming from `response.customError` method? Do we need to use it then?
As I can see we still use for general error handler siemResponse

```
        return siemResponse.error({
          body: error.message,
          statusCode: error.statusCode,
        });

```

Looks like we can end up with different error response structure depends on when error was thrown

Reply:
Reading the description

> Replaced the usage of response.custom for error response with response.customError to be consistent with other Kibana plugins. This introduces a slight breaking change as response.customError doesn't pass the status code to the response body.

So this is not entirely true, because as far as I can see from the response:

- the body contains `statusCode`
- also, the body now contains the new `error` property

![](https://puu.sh/IJwtJ/99c71fc3f9.png)

I'd suggest not to introduce this `error` property yet, and so not use the `response.customError`. Let's figure out the problem with `siemResponse` etc later.

Reply:
Also please use the underscore case: it should be `status_code` and not `statusCode`

Reply:
> the body contains statusCode
> also, the body now contains the new error property

It turns out these properties are added automatically by the Kibana server when `response.customError` is returned from a route handler. It's interesting because our test didn't show these properties in the error response body. Probably, they are added somewhere deep inside the server before returning a response. So unit test doesn't capture them.

But okay, let's revert the error response related changes for not and figure out how to migrate away from `SiemResponseFactory` later.